### DISCUSSION TITLE
[Built-in commands (e.g ls, ps, etc)] to return [structured table data].

### DISCUSSION THREAD

**Discussion by good-unit** [xonsh/xonsh#5129] [Ideas]
To date commands like [ls, ps, netstat, etc] return a [wall of text / big string] which needs to be parsed by tools like awk, etc -- before becoming useful to process.

NuShell built-in commands (https://www.nushell.sh/, https://www.nushell.sh/book/quick_tour.html):
  * Return: tables (who's columns have types (int, date, etc), and on which SQL type queries are possible).
  * Have Conditions (built-in commands): where, sort-by (used with piping).
  * Get/select columns (typed list returned)

This means there is no need for specialized "find" type of commands (for each tool like: ls)... queries (where, sort-by, etc) work on all [tables / columns / types] returned by all commands (not just ls -- but also ps, sys, netstat etc).

Has this idea been considered for Xonsh (as built-ins or plug-ins)?

You may ask... why don't I just switch to NuShell and use that:
* NuShells downside is that its language is very static (compiled) and doesn't have the dynamic abilities like @() and the power of python libraries behind it.

**Comment by anki-code** [comment]
I believe it's possible to implement this as [xontrib](https://github.com/topics/xontrib). You can create `where` and `sort-by` aliases and implement something like `ls | table | sort-by created-at`. Also take a look into [xontrib-pipeliner](https://github.com/anki-code/xontrib-pipeliner).
